File: proposed-projects/round-trip-experiment/reverse/20260909T142951Z_4ff7cdec/20260910T082818Z_6d83d454/ported_implementation/gbnf/rules_builder/symbol_ids.py
from typing import Dict, Iterator, Optional, Tuple


class SymbolIds:
    def __init__(self) -> None:
        self._map: Dict[str, int] = {}
        self._pos: Dict[str, int] = {}
        self._reverse_map: Dict[int, str] = {}

    def entries(self) -> Iterator[Tuple[str, int]]:
        return iter(list(self._map.items()))

    def __iter__(self) -> Iterator[Tuple[str, int]]:
        return self.entries()

    def __len__(self) -> int:
        return len(self._map)

    @property
    def size(self) -> int:
        return len(self._map)

    def get(self, key: str) -> Optional[int]:
        return self._map.get(key)

    def set(self, key: str, value: int, pos: int) -> None:
        self._map[key] = value
        self._reverse_map[value] = key
        self._pos[key] = pos

    def has(self, key: str) -> bool:
        return key in self._map

    def reverse_get(self, key: int) -> str:
        value = self._reverse_map.get(key)
        if value is None:
            raise ValueError(f"SymbolIds does not contain value: {key}")
        return value

    def get_pos(self, key: str) -> int:
        value = self._pos.get(key)
        if value is None:
            raise ValueError(f"SymbolIds does not contain key: {key}")
        return value

File: packages/gbnf-experiment/data/20260909T151143Z_170b737d/ported_implementation/src/rules_builder/symbol_ids.ts
import { KeyError } from "../utils/python_compat.ts";

export class SymbolIds {
  private __map__: Map<string, number>;
  private __pos__: Map<string, number>;
  private __reverse_map__: Map<number, string>;

  constructor() {
    this.__map__ = new Map();
    this.__pos__ = new Map();
    this.__reverse_map__ = new Map();
  }

  items(): IterableIterator<[string, number]> {
    return this.__map__.entries();
  }

  [Symbol.iterator](): IterableIterator<[string, number]> {
    return this.__map__.entries();
  }

  get length(): number {
    return this.__map__.size;
  }

  /** Equivalent of `symbol_ids[key]`; raises like Python's `dict.__getitem__`. */
  getItem(key: string): number {
    const value = this.__map__.get(key);
    if (value === undefined) {
      throw new KeyError(key);
    }
    return value;
  }

  set(key: string, value: number, pos: number): void {
    this.__map__.set(key, value);
    this.__reverse_map__.set(value, key);
    this.__pos__.set(key, pos);
  }

  has(key: string): boolean {
    return this.__map__.has(key);
  }

  reverse_get(key: number): string {
    const val = this.__reverse_map__.get(key);
    if (val === undefined) {
      throw new Error(`SymbolIds does not contain value: ${key}`);
    }
    return val;
  }

  get_pos(key: string): number {
    const val = this.__pos__.get(key);
    if (val === undefined) {
      throw new Error(`SymbolIds does not contain key: ${key}`);
    }
    return val;
  }
}
